### ISSUE TITLE
Rule Proposal: `vue/external-component-files

### ISSUE THREAD

**Issue by Snugug** [vuejs/eslint-plugin-vue#192]
**Please describe what the rule should do:**

Enforce external files for Single File Components. Config would have `template`, `style` and `script` options, and would have `require`, or`disallow`.

**What category of rule is this?**

- [x] Enforces Code Style
- [ ] Warns about a potential error
- [x] Suggests an alternative way of doing something
- [ ] Other (please specify)

**Provide 2-3 code examples that this rule will warn about**

**Require external template files**
```vue
<!-- This is OK -->
<template src="./foo.html"></template>

<!-- This is an Error -->
<template>
  <div>Foo</div>
</template>
```

**Disallow external style files**
```vue
<!-- This is OK -->
<style>
  .foo {
    color: red;
  }
</style>

<!-- This is an Error -->
<style src="./foo.css"></style>
```

**Comment by piehei** [comment]
I've been using SFCs for quite a while but haven't come up with a need for external templates. I'm not saying this wouldn't be useful, just that I haven't needed anything like this.

Why would you want to make the templating external? Then you lose the coupling to your component spesific logic.

Just interested :)

**Comment by Snugug** [comment]
I have no need for external templates (external CSS and JS, I've run in to instances of, but not templates). I'd most like to have that option to disallow external templates because there are people on my team who have done this that I'd like to prevent from doing this. That said, I like giving people options when providing linting rules, so give people the option to if they'd like!

**Comment by piehei** [comment]
To familiarize myself with the ESLint, I've started to work on implementing this new rule. I will make a PR when it's ready.

Should this be split in three (no-external-style, no-external-template, no-external-script) or stay as a single linting rule?

**Comment by mysticatea** [comment]
I'd like to separate the rule to 3:

- `external-script`
- `external-style`
- `external-template`

And each rule has an option `"never"`/`"always"`.

**Comment by piehei** [comment]
That sounds good. This will be ready during the weekend.

**Comment by michalsnik** [comment]
I'm worrying about one thing here, it probably won't be possible to disable those rules per file, so if I usually write everything in single-file components as recommended, but have one component that because of it's nature has quite big template that I want to put it in separate file I'll have to disable the rule entirely.

Can you please confirm or deny my concerns @mysticatea ? You were working on vue-eslint-parser so you'll better know if we can catch `eslint` comments from the root of the `.vue` AST and make `eslint` respect them.

**Comment by piehei** [comment]
That's a fair point @michalsnik. If eslint-disable-bla-bla comments from the root of the SFC can be detected, then everything's fine I guess. If not, should there be exception filename configuration in addition to the never/always? Or any other workaround?

**Comment by michalsnik** [comment]
Separate file with paths to files that should be ignored regarding one rule seems like an extra job that nobody would like to take care about. If it won't be possible to detect the root comments, then I think we should consider dropping this idea and simply not do it.

**Comment by mysticatea** [comment]
Currently, `vue-eslint-parser` does not support `<!-- eslint-disable -->` directive comments in HTML comments. People have to write `/* eslint external-template: off */` in the `<script>` element.

I'm not sure it's possible.
If it merges the `<!-- eslint-disable -->` comments into the `programNode.comments` list, it can cause confusion to core rules about comments. And ESLint does not expose the API to control disabling area.

**Comment by michalsnik** [comment]
I've been thinking about this once again, and I don't think we should add such rules. They're not possible to both implement and make easy to configure. Plus this proposition is not aligned with official styleguide.
We should enforce usage of single file components, and create small components instead of moving logic or template to separate files and in the result making them too big.
